enhancements/machine-api/converting-machine-api-to-cluster-api.md

```diff
@@ -0,0 +1,782 @@
+However, with the above described workflow, the user is not presented with a synchronous response to the migration request.
+
+To improve the feedback loop, a webhook will be introduced to validate the migration request.
+
+The migration request will always be persisted, however, the webhook will check for the presence of the sync controller's failure condition, and return a warning response to the user if the migration is not possible.
+
+The users intention to migrate will be persisted, and the sync controller will continue to attempt to migrate the resource until the condition is resolved, but, the user is also informed of the reason for the failure synchronously.
+
+This extension is not required for the minimum viable version of this project, but, will be added in a future iteration, prior to the GA release.
+
+### API Extensions
+
+This enhancement introduces a new `authoritativeAPI` field on `spec` and `status` of Machine API resources.
+Users will set the value of this field to determine which of the Machine API or Cluster API controllers will own the resource.
+
+The `status` will also include a `synchronizedGeneration` field to indicate the generation of the authoritative resource that the non-authoritative resource is synchronised with.
+
+Machine API conditions will also include a new `Synchronized` condition type, which will be used to indicate when the non-authoritative resource is fully synchronised with the authoritative resource.
+
+#### Authority of resources and controller reconciliations
+
+To ensure that only one controller acts on any resource (Machine, MachineSet etc) at any one time,
+a new field `status.authoritative-api` will be implemented.
+The new field will be applied to each Machine API resource and will determine which controller is authoritative for the resource.
+
+The accepted values for the field will be either `machine.openshift.io` or `cluster.x-k8s.io` (the respective API groups for Machine API and Cluster API).
+If the field is not present, it will be added by the sync controller and defaulted to `machine.openshift.io`.
+
+When the Machine API controller attempts to reconcile an object, it must first check for the presence of the field.
+If it is not present, then the Machine API controller should reconcile the object.
+It is expected that the sync controller will default it to `machine.openshift.io` in this case.
+
+We will implement predicates/event filters for both Machine API and Cluster API controllers.
+
+Machine API controllers will be allowed to reconcile any object that is either:
+
+* Missing the authoritative field in the status (as today)
+* Has the authoritative field set with the `machine.openshift.io` value
+
+Cluster API controllers will be allowed to reconcile any object that either:
+
+* Has a Machine API equivalent AND has the authoritative field set with the `cluster.x-k8s.io` value
```

Comment:
How would the Machine controller know this?

Reply:
Updated now, the sync controller will use the pause mechanisms in Cluster API to control this, similar mechanisms will be added to Machine API as well